Release 4.5.2 to address CVE-2021-44228

We should release 4.5.2 version to quickly address CVE-2021-44228.
The PR #1865 is prerequisite.

Note: there is no known exploit using spotbugs possible yet, but since we ship and use log4j in affected version, and there are web services that run spotbugs as a part of the build pipeline, an attacker could manage to craft a pull request that contains the Java code that would somehow manage the bad pattern be part of the message logged by spotbugs.

@KengoTODA : the build fails with this error below: any idea?
```
> Task :closeSonatypeStagingRepository FAILED
FAILURE: Build failed with an exception.


Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0.
* What went wrong:

Execution failed for task ':closeSonatypeStagingRepository'.
You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.
> Staging repository is still transitioning after defined time. Consider its increment. StagingRepository(id=comgithubspotbugs-1253, state=open, transitioning=true)
```

--- a/.github/dependabot.yml
+++ b/.github/dependabot.yml
@@ -4,3 +4,7 @@ updates:
     directory: "/"
     schedule:
       interval: "daily"
+  - package-ecosystem: "gradle"
+    directory: "/buildSrc"
+    schedule:
+      interval: "daily"

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -18,6 +18,7 @@ jobs:
           java-version: '17'
           distribution: temurin
           cache: gradle
+      - uses: gradle/wrapper-validation-action@v1
       - name: Download Eclipse
         run: |
           wget -nv 'https://www.eclipse.org/downloads/download.php?file=/eclipse/downloads/drops4/R-4.6.3-201703010400/eclipse-SDK-4.6.3-linux-gtk-x86_64.tar.gz&r=1' -O eclipse-SDK-4.6.3-linux-gtk-x86_64.tar.gz
@@ -31,7 +32,7 @@ jobs:
           if [ "$GPG_SECRET_PASSPHRASE" != "" ]; then
             gpg --quiet --batch --yes --decrypt --passphrase="$GPG_SECRET_PASSPHRASE" --output spotbugs.jks .github/workflows/spotbugs.jks.gpg;
           fi
-          ./gradlew spotlessCheck build smoketest ${SONAR_LOGIN:+sonarqube} --no-daemon -Dsonar.login=$SONAR_LOGIN
+          ./gradlew spotlessCheck build smoketest ${SONAR_LOGIN:+sonarqube} --no-daemon -Dsonar.login=$SONAR_LOGIN --scan
         env:
           GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
           GPG_SECRET_PASSPHRASE: ${{ secrets.GPG_SECRET_PASSPHRASE }}

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -17,6 +17,7 @@ jobs:
           java-version: '17'
           distribution: temurin
           cache: gradle
+      - uses: gradle/wrapper-validation-action@v1
       - name: Download Eclipse
         run: |
           wget 'https://www.eclipse.org/downloads/download.php?file=/eclipse/downloads/drops4/R-4.6.3-201703010400/eclipse-SDK-4.6.3-linux-gtk-x86_64.tar.gz&r=1' -O eclipse-SDK-4.6.3-linux-gtk-x86_64.tar.gz

--- a/build.gradle
+++ b/build.gradle
@@ -13,6 +13,7 @@ apply from: "$rootDir/gradle/java.gradle"
 apply from: "$rootDir/gradle/jacoco.gradle"
 
 subprojects {
+  apply plugin: "constraints"
   apply from: "$rootDir/gradle/java.gradle"
   apply from: "$rootDir/gradle/eclipse.gradle"
   apply from: "$rootDir/gradle/idea.gradle"

--- a/buildSrc/build.gradle.kts
+++ b/buildSrc/build.gradle.kts
@@ -1,0 +1,10 @@
+plugins {
+    `kotlin-dsl`
+}
+
+repositories {
+    gradlePluginPortal()
+}
+dependencies {
+    implementation("com.diffplug.gradle:goomph:3.34.1")
+}

--- a/buildSrc/src/main/kotlin/constraints.gradle.kts
+++ b/buildSrc/src/main/kotlin/constraints.gradle.kts
@@ -1,0 +1,22 @@
+plugins {
+    `java-library`
+}
+
+dependencies {
+    constraints {
+        implementation("org.apache.logging.log4j:log4j-core") {
+            version {
+                strictly("[2.17.1, 3[")
+                prefer("2.17.1")
+            }
+            because("CVE-2021-44228, CVE-2021-45046, CVE-2021-45105, CVE-2021-44832: Log4j vulnerable to remote code execution and other critical security vulnerabilities")
+        }
+        implementation("ch.qos.logback:logback-core") {
+            version {
+                strictly("[1.2.9, 2[")
+                prefer("1.2.10")
+            }
+            because("CVE-2021-42550: Logback vulnerable to remote code execution vulnerabilities")
+        }
+    }
+}

--- a/buildSrc/src/main/kotlin/eclipse-convention.gradle.kts
+++ b/buildSrc/src/main/kotlin/eclipse-convention.gradle.kts
@@ -7,7 +7,7 @@ val pdeTool by configurations.creating {
 }
 
 eclipseMavenCentral {
-  release("4.6.3") {
+  release("4.12.0") {
     compileOnly("org.eclipse.ant.core")
     compileOnly("org.eclipse.core.resources")
     compileOnly("org.eclipse.core.runtime")

--- a/buildSrc/src/main/kotlin/eclipse-convention.gradle.kts
+++ b/buildSrc/src/main/kotlin/eclipse-convention.gradle.kts
@@ -1,0 +1,46 @@
+plugins {
+  id ("com.diffplug.eclipse.mavencentral")
+}
+
+val pdeTool by configurations.creating {
+  setTransitive(false)
+}
+
+eclipseMavenCentral {
+  release("4.6.3") {
+    compileOnly("org.eclipse.ant.core")
+    compileOnly("org.eclipse.core.resources")
+    compileOnly("org.eclipse.core.runtime")
+    compileOnly("org.eclipse.jdt.core")
+    compileOnly("org.eclipse.jdt.ui")
+    compileOnly("org.eclipse.jface")
+    compileOnly("org.eclipse.pde")
+    compileOnly("org.eclipse.ui.workbench")
+    testImplementation("org.eclipse.core.runtime")
+
+    dep("pdeTool", "org.eclipse.pde.build")
+
+    // TODO these packages are not listed in the manifest
+    compileOnly("org.eclipse.pde.ui")
+    compileOnly("org.eclipse.swt")
+
+    // necessary to build with the org.eclipse.swt module
+    useNativesForRunningPlatform()
+
+    constrainTransitivesToThisRelease()
+  }
+}
+
+/**
+ * Unzip "org.eclipse.pde.build" package into the outputDir.
+ */
+val unzipPdeTool = tasks.register<Copy>("unzipPdeTool") {
+  from(zipTree(pdeTool.singleFile))
+  into("$buildDir/pdeTool")
+}
+
+dependencies {
+  compileOnly(files("$buildDir/pdeTool/pdebuild.jar"){
+    builtBy(unzipPdeTool)
+  })
+}

--- a/eclipsePlugin-junit/build.gradle
+++ b/eclipsePlugin-junit/build.gradle
@@ -5,6 +5,10 @@ plugins {
 apply from: "$rootDir/gradle/checkstyle.gradle"
 apply from: "$rootDir/gradle/jacoco.gradle"
 
+tasks.named('compileJava', JavaCompile).configure {
+  options.release = 11
+}
+
 dependencies {
   implementation project(':eclipsePlugin')
   testImplementation 'junit:junit:4.13.2'

--- a/eclipsePlugin-junit/build.gradle
+++ b/eclipsePlugin-junit/build.gradle
@@ -1,33 +1,14 @@
+plugins {
+  id "eclipse-convention"
+}
+
 apply from: "$rootDir/gradle/checkstyle.gradle"
 apply from: "$rootDir/gradle/jacoco.gradle"
-
-def localProps = new Properties()
-localProps.load(new FileInputStream("$rootDir/eclipsePlugin/local.properties"))
-
-def eclipseRootDir = file(localProps.getProperty('eclipseRoot.dir'))
-// Eclipse 4.5+ uses a different directory layout under macOS. Try to detect this first.
-def eclipseExecutable = new File(eclipseRootDir, "Contents/MacOS/eclipse")
-def eclipsePluginsDir = new File(eclipseRootDir, "Contents/Eclipse/plugins")
-if (!eclipseExecutable.exists()) {
-  // Fall back to non-macOS directory layout.
-  eclipsePluginsDir = new File(eclipseRootDir, "plugins")
-}
 
 dependencies {
   implementation project(':eclipsePlugin')
   testImplementation 'junit:junit:4.13.2'
   testImplementation 'org.mockito:mockito-core:4.2.0'
-
-  // List below includes all Eclipse SDK plugins except few causing troubles.
-  // TODO: it should include what is required in the MANIFEST.MF, and nothing else
-  testImplementation fileTree(dir:eclipsePluginsDir, include:'**/*.jar',
-          exclude:[
-                  '**/datanucleus-enhancer*.jar',
-                  'edu.umd.cs.findbugs.**/*.jar',
-                  'org.mockito*.jar',
-                  'com.github.spotbugs.**/*.jar',
-                  '**/*source_*.jar'
-          ])
 }
 
 tasks.named('jacocoTestReport', JacocoReport).configure {

--- a/eclipsePlugin-test/build.gradle
+++ b/eclipsePlugin-test/build.gradle
@@ -1,16 +1,7 @@
-apply from: "$rootDir/gradle/checkstyle.gradle"
-
-def localProps = new Properties()
-localProps.load(new FileInputStream("$rootDir/eclipsePlugin/local.properties"))
-
-def eclipseRootDir = file(localProps.getProperty('eclipseRoot.dir'))
-// Eclipse 4.5+ uses a different directory layout under macOS. Try to detect this first.
-def eclipseExecutable = new File(eclipseRootDir, "Contents/MacOS/eclipse")
-def eclipsePluginsDir = new File(eclipseRootDir, "Contents/Eclipse/plugins")
-if (!eclipseExecutable.exists()) {
-  // Fall back to non-macOS directory layout.
-  eclipsePluginsDir = new File(eclipseRootDir, "plugins")
+plugins {
+  id "eclipse-convention"
 }
+apply from: "$rootDir/gradle/checkstyle.gradle"
 
 sourceSets {
   main {
@@ -29,16 +20,6 @@ dependencies {
   implementation(project(':eclipsePlugin')) {
     transitive = true
   }
-  // List below includes all Eclipse SDK plugins except few causing troubles.
-  // TODO: it should include what is required in the MANIFEST.MF, and nothing else
-  compileOnly fileTree(dir:eclipsePluginsDir, include:'**/*.jar',
-          exclude:[
-                  '**/datanucleus-enhancer*.jar',
-                  'edu.umd.cs.findbugs.**/*.jar',
-                  'org.mockito*.jar',
-                  'com.github.spotbugs.**/*.jar',
-                  '**/*source_*.jar'
-          ])
 
   implementation project(':test-harness')
   implementation 'org.hamcrest:hamcrest-all:1.3'

--- a/eclipsePlugin-test/build.gradle
+++ b/eclipsePlugin-test/build.gradle
@@ -3,6 +3,10 @@ plugins {
 }
 apply from: "$rootDir/gradle/checkstyle.gradle"
 
+tasks.named('compileJava', JavaCompile).configure {
+  options.release = 11
+}
+
 sourceSets {
   main {
     java {

--- a/eclipsePlugin/build.gradle
+++ b/eclipsePlugin/build.gradle
@@ -4,6 +4,7 @@ import static groovy.io.FileType.FILES
 // apply from: "$rootDir/gradle/maven.gradle"
 plugins {
   id 'org.ajoberstar.grgit' version '4.1.1'
+  id "eclipse-convention"
 }
 
 def readLastCommitHash() {
@@ -41,33 +42,27 @@ sourceSets {
   }
 }
 
-def eclipsePluginId = 'com.github.spotbugs.plugin.eclipse'
+def findEclipseExecutable() {
+  if (!file("$projectDir/local.properties").exists()) {
+    return null
+  }
 
-def localProps = new Properties()
-localProps.load(new FileInputStream("$projectDir/local.properties"))
-
-def eclipseRootDir = file(localProps.getProperty('eclipseRoot.dir'))
-// Eclipse 4.5+ uses a different directory layout under macOS. Try to detect this first.
-def eclipseExecutable = new File(eclipseRootDir, "Contents/MacOS/eclipse")
-def eclipsePluginsDir = new File(eclipseRootDir, "Contents/Eclipse/plugins")
-if (!eclipseExecutable.exists()) {
-   // Fall back to non-macOS directory layout.
-   eclipseExecutable = new File(eclipseRootDir, "eclipse")
-   eclipsePluginsDir = new File(eclipseRootDir, "plugins")
+  def localProps = new Properties()
+  localProps.load(new FileInputStream("$projectDir/local.properties"))
+  def eclipseRootDir = file(localProps.getProperty('eclipseRoot.dir'))
+  // Eclipse 4.5+ uses a different directory layout under macOS. Try to detect this first.
+  def eclipseExecutable = new File(eclipseRootDir, "Contents/MacOS/eclipse")
+  if (!eclipseExecutable.exists()) {
+    // Fall back to non-macOS directory layout.
+    eclipseExecutable = new File(eclipseRootDir, "eclipse")
+  }
+  return eclipseExecutable
 }
 
-dependencies {
-  // List below includes all Eclipse SDK plugins except few causing troubles.
-  // TODO: it should include what is required in the MANIFEST.MF, and nothing else
-  compileOnly fileTree(dir:eclipsePluginsDir, include:'**/*.jar',
-    exclude:[
-        '**/datanucleus-enhancer*.jar',
-        'edu.umd.cs.findbugs.**/*.jar',
-        'org.mockito*.jar',
-        'com.github.spotbugs.**/*.jar',
-        '**/*source_*.jar'
-        ])
+def eclipsePluginId = 'com.github.spotbugs.plugin.eclipse'
+def eclipseExecutable = findEclipseExecutable()
 
+dependencies {
   api (project(':spotbugs')) {
     transitive = true
   }
@@ -92,19 +87,7 @@ tasks.named('javadoc', Javadoc).configure {
 }
 
 def copyLibsForEclipse = tasks.register('copyLibsForEclipse', Copy) {
-    from ('../spotbugs/.libs') {
-      include "*.jar"
-      exclude "*xml-apis*.jar"
-      exclude "*ant-*.jar"
-      exclude "Apple*.jar"
-      exclude "hamcrest*.jar"
-      exclude "slf4j-*alpha*.jar"
-      exclude "Saxon*.jar"
-    }
-    from ('../spotbugs/build/libs') {
-      include "*.jar"
-    }
-    from ('../spotbugs-annotations/build/libs') {
+    from (configurations.runtimeClasspath) {
       include "*.jar"
     }
     into "lib"
@@ -250,7 +233,7 @@ def testPluginJar = tasks.register('testPluginJar') {
             .matching { include 'lib/spotbugs.jar' }
             .singleFile
     if (!spotbugsJar.exists()) {
-      throw new GradleException('Eclipse plugin does not contain spotbugs.jar');
+      throw new TaskInstantiationException('Eclipse plugin does not contain spotbugs.jar');
     } else {
       println 'Eclipse plugin contains spotbugs.jar'
     }
@@ -465,6 +448,14 @@ ext.signJar = { File dir ->
   }
 }
 
+def confirmEclipse = tasks.register('confirmEclipse') {
+  doFirst {
+    if (eclipseExecutable == null || !eclipseExecutable.exists()) {
+      throw new GradleScriptException("No eclipse found in local")
+    }
+  }
+}
+
 def generateP2Metadata = tasks.register('generateP2Metadata', Exec) {
   doFirst {
     project.delete "${buildDir}/site/eclipse/artifacts.xml"
@@ -472,7 +463,7 @@ def generateP2Metadata = tasks.register('generateP2Metadata', Exec) {
     signJar(file("${buildDir}/site/eclipse"))
   }
   inputs.file 'local.properties'
-  dependsOn pluginJar, featureJar, siteXml, siteHtml
+  dependsOn confirmEclipse, pluginJar, featureJar, siteXml, siteHtml
   commandLine "${eclipseExecutable}", '-nosplash',
     '-application', 'org.eclipse.equinox.p2.publisher.UpdateSitePublisher',
     '-metadataRepository', "file:${buildDir}/site/eclipse", // TODO : May not work on Windows
@@ -487,7 +478,7 @@ def generateCandidateP2Metadata = tasks.register('generateCandidateP2Metadata', 
     signJar(file("${buildDir}/site/eclipse-candidate"))
   }
   inputs.file 'local.properties'
-  dependsOn pluginCandidateJar, featureCandidateJar, siteCandidateXml, siteCandidateHtml
+  dependsOn confirmEclipse, pluginCandidateJar, featureCandidateJar, siteCandidateXml, siteCandidateHtml
   commandLine "${eclipseExecutable}", '-nosplash',
     '-application', 'org.eclipse.equinox.p2.publisher.UpdateSitePublisher',
     '-metadataRepository', "file:${buildDir}/site/eclipse-candidate", // TODO : May not work on Windows
@@ -502,7 +493,7 @@ def generateP2MetadataDaily = tasks.register('generateP2MetadataDaily', Exec) {
     signJar(file("${buildDir}/site/eclipse-daily"))
   }
   inputs.file 'local.properties'
-  dependsOn pluginDailyJar, featureDailyJar, siteDailyXml, siteDailyHtml
+  dependsOn confirmEclipse, pluginDailyJar, featureDailyJar, siteDailyXml, siteDailyHtml
   commandLine "${eclipseExecutable}", '-nosplash',
     '-application', 'org.eclipse.equinox.p2.publisher.UpdateSitePublisher',
     '-metadataRepository', "file:${buildDir}/site/eclipse-daily", // TODO : May not work on Windows
@@ -517,7 +508,7 @@ def generateP2MetadataStableLatest = tasks.register('generateP2MetadataStableLat
     signJar(file("${buildDir}/site/eclipse-stable-latest"))
   }
   inputs.file 'local.properties'
-  dependsOn pluginStableLatestJar, featureStableLatestJar, siteStableLatestXml, siteStableLatestHtml
+  dependsOn confirmEclipse, pluginStableLatestJar, featureStableLatestJar, siteStableLatestXml, siteStableLatestHtml
   commandLine "${eclipseExecutable}", '-nosplash',
     '-application', 'org.eclipse.equinox.p2.publisher.UpdateSitePublisher',
     '-metadataRepository', "file:${buildDir}/site/eclipse-stable-latest", // TODO : May not work on Windows
@@ -536,7 +527,11 @@ def eclipseSiteZip = tasks.register('eclipseSiteZip', Zip) {
   archiveFileName = "eclipsePlugin.zip"
 }
 
-tasks.named('assemble').configure { dependsOn eclipseSite, eclipseSiteZip }
+if (eclipseExecutable == null) {
+  println "Create local.properties to build eclipse site"
+} else {
+  tasks.named('assemble').configure { dependsOn eclipseSiteZip }
+}
 
 spotbugs {
   ignoreFailures = true

--- a/eclipsePlugin/build.gradle
+++ b/eclipsePlugin/build.gradle
@@ -89,6 +89,7 @@ tasks.named('javadoc', Javadoc).configure {
 def copyLibsForEclipse = tasks.register('copyLibsForEclipse', Copy) {
     from (configurations.runtimeClasspath) {
       include "*.jar"
+      exclude "Saxon*.jar"
     }
     into "lib"
 }

--- a/eclipsePlugin/build.gradle
+++ b/eclipsePlugin/build.gradle
@@ -7,6 +7,15 @@ plugins {
   id "eclipse-convention"
 }
 
+// Force Eclipse use Java 1.8, otherwise it will get Java 17 (!) from gradle
+eclipse {
+  jdt {
+    sourceCompatibility = 8
+    targetCompatibility = 8
+    javaRuntimeName = "JavaSE-1.8"
+  }
+}
+
 def readLastCommitHash() {
   // Same as 'git log -1 --pretty=format:"%h"'.execute().getText()
   org.ajoberstar.grgit.Grgit.open(dir: rootDir).head().abbreviatedId
@@ -55,6 +64,10 @@ def findEclipseExecutable() {
   if (!eclipseExecutable.exists()) {
     // Fall back to non-macOS directory layout.
     eclipseExecutable = new File(eclipseRootDir, "eclipse")
+  }
+  logger.warn('found eclipseExecutable at ' + eclipseExecutable)
+  if (!eclipseExecutable.exists()) {
+    logger.error('eclipseExecutable at ' + eclipseExecutable + ' does not exist!')
   }
   return eclipseExecutable
 }
@@ -452,7 +465,8 @@ ext.signJar = { File dir ->
 def confirmEclipse = tasks.register('confirmEclipse') {
   doFirst {
     if (eclipseExecutable == null || !eclipseExecutable.exists()) {
-      throw new GradleScriptException("No eclipse found in local")
+      logger.warn('found eclipseExecutable at 2 ' + eclipseExecutable)
+      throw new GradleScriptException("No eclipse found in local", null)
     }
   }
 }

--- a/eclipsePlugin/build.gradle
+++ b/eclipsePlugin/build.gradle
@@ -58,11 +58,19 @@ def findEclipseExecutable() {
 
   def localProps = new Properties()
   localProps.load(new FileInputStream("$projectDir/local.properties"))
-  def eclipseRootDir = file(localProps.getProperty('eclipseRoot.dir'))
+  def eclipseRootDir = new File(localProps.getProperty('eclipseRoot.dir'))
   // Eclipse 4.5+ uses a different directory layout under macOS. Try to detect this first.
-  def eclipseExecutable = new File(eclipseRootDir, "Contents/MacOS/eclipse")
-  if (!eclipseExecutable.exists()) {
-    // Fall back to non-macOS directory layout.
+  OperatingSystem os = org.gradle.nativeplatform.platform.internal.DefaultNativePlatform.currentOperatingSystem;
+  def eclipseExecutable;
+  if(os.isMacOsX()){
+    eclipseExecutable = new File(eclipseRootDir, "Contents/MacOS/eclipse")
+    if (!eclipseExecutable.exists()) {
+      // Fall back to non-macOS directory layout.
+      eclipseExecutable = new File(eclipseRootDir, "eclipse")
+    }
+  } else if (os.isWindows()) {
+    eclipseExecutable = new File(eclipseRootDir, "eclipse.exe")
+  } else {
     eclipseExecutable = new File(eclipseRootDir, "eclipse")
   }
   logger.warn('found eclipseExecutable at ' + eclipseExecutable)

--- a/settings.gradle
+++ b/settings.gradle
@@ -2,15 +2,17 @@ plugins {
   id "com.gradle.enterprise" version "3.8"
 }
 
-def eclipseFile = "eclipsePlugin/local.properties"
-if (new File(rootDir, eclipseFile).exists()) {
-  include ':eclipsePlugin', ':eclipsePlugin-test', ':eclipsePlugin-junit', ':spotbugsTestCases', ':spotbugs', ':spotbugs-tests', ':test-harness', ':spotbugs-annotations', ':spotbugs-ant'
-} else {
-  logger.lifecycle('Eclipse plugin configuration (' + eclipseFile + ') was not found. Skipping Eclipse plugin...')
-  include ':spotbugsTestCases', ':spotbugs', ':spotbugs-tests', ':test-harness', ':spotbugs-annotations', ':spotbugs-ant'
-}
-
-include ':test-harness-core', ':test-harness-jupiter'
+include ':eclipsePlugin'
+include ':eclipsePlugin-test'
+include ':eclipsePlugin-junit'
+include ':spotbugs'
+include ':spotbugs-annotations'
+include ':spotbugs-ant'
+include ':spotbugs-tests'
+include ':spotbugsTestCases'
+include ':test-harness'
+include ':test-harness-core'
+include ':test-harness-jupiter'
 
 gradleEnterprise {
   buildScan {

--- a/spotbugs-annotations/build.gradle
+++ b/spotbugs-annotations/build.gradle
@@ -2,6 +2,21 @@ apply from: "$rootDir/gradle/checkstyle.gradle"
 apply from: "$rootDir/gradle/javadoc.gradle"
 apply from: "$rootDir/gradle/maven.gradle"
 
+// This library is consumed by lot of clients
+// It should be compiled with lowest possible JLS level
+tasks.named('compileJava', JavaCompile).configure {
+  sourceCompatibility = 1.5
+  targetCompatibility = 1.5
+}
+
+eclipse {
+  jdt {
+    sourceCompatibility = 1.5
+    targetCompatibility = 1.5
+    javaRuntimeName = "J2SE-1.5"
+  }
+}
+
 dependencies {
   api 'com.google.code.findbugs:jsr305:3.0.2'
 }

--- a/spotbugs/build.gradle
+++ b/spotbugs/build.gradle
@@ -316,6 +316,26 @@ apply from: "$rootDir/gradle/jigsaw.gradle"
 tasks.named('spotbugsTest').configure {
   ignoreFailures = true
 }
+
+dependencies {
+  constraints {
+    logBinding("org.apache.logging.log4j:log4j-core") {
+      version {
+        strictly("[2.17.1, 3[")
+        prefer("2.17.1")
+      }
+      because("CVE-2021-44228, CVE-2021-45046, CVE-2021-45105, CVE-2021-44832: Log4j vulnerable to remote code execution and other critical security vulnerabilities")
+    }
+    logBinding("ch.qos.logback:logback-core") {
+      version {
+        strictly("[1.2.9, 2[")
+        prefer("1.2.10")
+      }
+      because("CVE-2021-42550: Logback vulnerable to remote code execution vulnerabilities")
+    }
+  }
+}
+
 // TODO : generatemanual (we should decide what to do with the manual)
 // TODO : generatepdfmanual
 // TODO : bugdesc

--- a/spotbugs/build.gradle
+++ b/spotbugs/build.gradle
@@ -2,6 +2,19 @@ apply from: "$rootDir/gradle/maven.gradle"
 apply from: "$rootDir/gradle/checkstyle.gradle"
 apply from: "$rootDir/gradle/javadoc.gradle"
 
+tasks.named('compileJava', JavaCompile).configure {
+    options.release = 8
+}
+
+// Force Eclipse use Java 1.8, otherwise it will get Java 17 (!) from gradle
+eclipse {
+  jdt {
+    sourceCompatibility = 8
+    targetCompatibility = 8
+    javaRuntimeName = "JavaSE-1.8"
+  }
+}
+
 configurations {
   // used only in distribution. It is not listed in pom.xml, so users like maven plugin don't use this dependency.
   logBinding
@@ -107,13 +120,11 @@ tasks.withType(Jar).configureEach {
   //destinationDir = file("$projectDir/lib")
 }
 
-eclipse.classpath {
-  plusConfigurations += [ configurations.appleExtensions ]
-}
-
 eclipse.classpath.file {
     whenMerged {
       classpath ->
+         classpath.entries.removeAll { entry -> entry.kind == 'lib' && entry.path.contains('build/classes/')}
+         classpath.entries.removeAll { entry -> entry.kind == 'lib' && entry.path.contains('build/resources/')}
          classpath.entries.removeAll { entry -> entry.kind == 'lib' && entry.path.contains('xml-apis')}
          classpath.entries.forEach {
              entry ->
@@ -122,20 +133,44 @@ eclipse.classpath.file {
                    entry.path = ".libs/" + java.nio.file.Paths.get(entry.path).getFileName().toString()
                    entry.exported = true
                }
-      }
+         }
+         Map entryByPath = classpath.entries.groupBy { entry -> entry.path }
+         entryByPath.each { key, values ->
+          if (values.size() > 1) {
+            def entry = values.first()
+            if (entry.kind == 'src') {
+              entry.includes = []
+              entry.excludes = []
+            }
+            int index = classpath.entries.indexOf entry
+            logger.lifecycle "Removing ${values.collect { it.path }}"
+            classpath.entries.removeAll values
+            logger.lifecycle "Adding ${entry.path}"
+            classpath.entries.add index, entry
+          }
+        }
     }
+}
+
+def copyLibsForEclipse = tasks.register('copyLibsForEclipse', Copy) {
+    from configurations.testCompileClasspath, configurations.testRuntimeClasspath, configurations.guiRuntimeClasspath, configurations.guiCompileClasspath, configurations.logBinding
+    into ".libs"
+    include "*.jar"
+    exclude "*xml-apis*.jar"
+    exclude "*spotbugs-annotations*.jar"
+    duplicatesStrategy = DuplicatesStrategy.EXCLUDE
 }
 
 def updateManifest = tasks.register('updateManifest') {
   inputs.file "$projectDir/META-INF/MANIFEST-TEMPLATE.MF"
   outputs.file "$projectDir/META-INF/MANIFEST.MF"
-  dependsOn configurations.runtimeClasspath
+  dependsOn configurations.runtimeClasspath, copyLibsForEclipse
   doLast {
     def manifestSpec = manifest {
       from "$projectDir/META-INF/MANIFEST-TEMPLATE.MF"
       attributes 'Main-Class': 'edu.umd.cs.findbugs.LaunchAppropriateUI',
                'Bundle-Version': project.version.replace('-', '.'),
-               'Bundle-ClassPath': 'spotbugs.jar,' +  configurations.runtimeClasspath.collect { it.name }.join(',')
+               'Bundle-ClassPath': 'spotbugs.jar,' +  fileTree(dir: '.libs').collect { projectDir.toPath().relativize(it.toPath()).toString() }.join(',')
     }
     // write manifests
     manifestSpec.writeTo("$projectDir/META-INF/MANIFEST.MF")

--- a/spotbugs/build.gradle
+++ b/spotbugs/build.gradle
@@ -126,25 +126,16 @@ eclipse.classpath.file {
     }
 }
 
-def copyLibsForEclipse = tasks.register('copyLibsForEclipse', Copy) {
-    from configurations.testCompileClasspath, configurations.testRuntimeClasspath, configurations.guiRuntimeClasspath, configurations.guiCompileClasspath, configurations.logBinding
-    into ".libs"
-    include "*.jar"
-    exclude "*xml-apis*.jar"
-    exclude "*spotbugs-annotations*.jar"
-    duplicatesStrategy = DuplicatesStrategy.EXCLUDE
-}
-
 def updateManifest = tasks.register('updateManifest') {
   inputs.file "$projectDir/META-INF/MANIFEST-TEMPLATE.MF"
   outputs.file "$projectDir/META-INF/MANIFEST.MF"
-  dependsOn configurations.runtimeClasspath, copyLibsForEclipse
+  dependsOn configurations.runtimeClasspath
   doLast {
     def manifestSpec = manifest {
       from "$projectDir/META-INF/MANIFEST-TEMPLATE.MF"
       attributes 'Main-Class': 'edu.umd.cs.findbugs.LaunchAppropriateUI',
                'Bundle-Version': project.version.replace('-', '.'),
-               'Bundle-ClassPath': 'spotbugs.jar,' +  fileTree(dir: '.libs').collect { projectDir.toPath().relativize(it.toPath()).toString() }.join(',')
+               'Bundle-ClassPath': 'spotbugs.jar,' +  configurations.runtimeClasspath.collect { it.name }.join(',')
     }
     // write manifests
     manifestSpec.writeTo("$projectDir/META-INF/MANIFEST.MF")

